daemon/pool.go

```diff
@@ -73,123 +100,445 @@ func NewDriverPool(factory FactoryFunction) *DriverPool {
-	for i := 0; i < n; i++ {
-		d, more := dp.queue.Get()
-		if !more {
-			return ErrPoolClosed.New()
+// killDriver stops are removes the driver from the queue.
+func (dp *DriverPool) killDriver(d Driver) {
+	dp.drivers.Lock()
+	delete(dp.drivers.all, d)
+	delete(dp.drivers.idle, d)
+	dp.running.Add(-1)
+	dp.exited.Add(1)
+	dp.drivers.Unlock()
+
+	if err := d.Stop(); err != nil {
+		dp.Logger.Errorf("error removing stopped driver: %s", err)
+	}
+}
+
+// scaleDiff returns current difference between the target number of instances and the
+// current number of running instances.
+func (dp *DriverPool) scaleDiff() int {
+	total := dp.running.Value()
+	return dp.target.Value() - total
+}
+
+// scale the driver pool to the current target number of instances.
+func (dp *DriverPool) scale() {
+	dn := dp.scaleDiff()
+	if dn == 0 {
+		return
+	}
+
+	// rescaleLater interrupts the scaling and serves the request first.
+	// It will make sure to continue scaling later.
+	rescaleLater := func(req driverRequest) {
+		select {
+		case dp.rescale <- struct{}{}:
+		default:
+		}
+		dp.waitOrScale(req)
+	}
+	if dn < 0 {
+		// scale down
+		for i := 0; i < -dn; i++ {
+			if d, ok := dp.peekIdle(); ok {
+				dp.killDriver(d)
+				continue
+			}
+			select {
+			case <-dp.stop:
+				return
+			case req := <-dp.get:
+				rescaleLater(req)
+				return
+			case d := <-dp.put:
+				dp.killDriver(d)
+			}
 		}
+		return
+	}
+	// scale up
+	for i := 0; i < dn; i++ {
+		select {
+		case req := <-dp.get:
+			rescaleLater(req)
+			return
+		case <-dp.stop:
+			return
+		case d := <-dp.put:
+			// received some existing instance
+			dp.setIdle(d)
+			i--
+			dn = dp.scaleDiff()
+		case dp.spawn <- struct{}{}:
+			// spawn request sent to the spawn goroutine
+			select {
+			case <-dp.stop:
+				return
+			case d := <-dp.put:
+				dp.setIdle(d)
+			case <-dp.spawnErr:
+				// ignore - already printed to the log
+			}
+		}
+	}
+}
 
-		if err := dp.remove(d); err != nil {
-			return err
+// waitOrScale is executed on the manager goroutine. It will either scale the pool up,
+// wait for an instance to become available, or scale the pool down. The function will
+// block until the request is served or cancelled.
+func (dp *DriverPool) waitOrScale(req driverRequest) {
+	// Note that we don't care about the pool closing in this function.
+	// If it happens, the client will receive this signal before we do and will
+	// cancel the request anyway. This reduces the number of select statements.
+
+	// returnDriver will either return the driver to the client,
+	// or will return it to the idle driver queue.
+	returnDriver := func(d Driver) {
+		select {
+		case <-req.cancel:
+			dp.setIdle(d)
+		case req.out <- d:
 		}
 	}
 
-	return nil
-}
+	// loop allows to re-evaluate scaling conditions
+	for {
+		dn := dp.scaleDiff()
+
+		if dp.running.Value()+dp.spawning.Value() == 0 {
+			if dn < 0 {
+				// This shouldn't really happen: pool is draining, but there are no running
```

Comment:
Do you mean "shouldn't" in the sense that it's not common, or in the sense that if it happens something is broken in the upstream logic? ("Uncommon" is fine; "broken" is more serious).

Reply:
This is more like a "broken" case - the scaling policy should not return values that lead to this state. I was more inclined to panic here, but it may be a bit too much (it will kill bblfshd instance).